quarkus-workshop-super-heroes/docs/src/docs/asciidoc/5a-contract-testing/happy-path-consumer-tests.adoc

```diff
@@ -0,0 +1,166 @@
+                        .stringType("name", DEFAULT_HERO_NAME)
+                        .integerType("level", DEFAULT_HERO_LEVEL)
+                        .stringType("picture", DEFAULT_HERO_PICTURE)
+                ).build()
+            )
+            .toPact(V4Pact.class);
+    }
+}
+----
+
+Here we are setting up a test to run with Pact and telling Pact what ports it should stand up the mock server on.
+We also define how the mock should behave.
+
+[WARNING]
+Because of the complexities of Kotlin classloading, Pact does not work with continuous testing.
+
+You will need to run tests the old-fashioned way, with
+
+[shell]
+----
+mvn verify
+----
+
+
+Run the build (using `mvn verify`).
+You should see a failure, because we have a mock, but nothing is using it.
+
+
+Add in a test method.
+You'll notice this is very similar to the one in `FightResourceTest`; you could even copy and paste from that class.
+
+[source,java]
+----
+     @Test
+    @PactTestFor(pactMethod = "randomHeroFoundPact")
+    void randomHeroFound() {
+        given()
+            .when().get("/api/fights/randomfighters")
+            .then()
+            .statusCode(OK.getStatusCode())
+            .contentType(APPLICATION_JSON)
+            .body("hero.name", Is.is(DEFAULT_HERO_NAME))
+            .body("hero.picture", Is.is(DEFAULT_HERO_PICTURE))
+            .body("hero.level", Is.is(DEFAULT_HERO_LEVEL))
+            .body("villain.name", Is.is(DefaultTestVillain.DEFAULT_VILLAIN_NAME))
+            .body("villain.picture", Is.is(DefaultTestVillain.DEFAULT_VILLAIN_PICTURE))
+            .body("villain.level", Is.is(DefaultTestVillain.DEFAULT_VILLAIN_LEVEL));
+    }
+----
+
+Run the tests.
+They should all pass.
+
+Note that we're only testing heroes here. If we were to test getting villains, the test would fail. Can you think why that might be?
+
+[TIP]
+When we use Pact, the Pact framework stands up a server which acts as a mock. Remember that we defined a global `@Mock` for the `VillainProxy`.
+That mock would interfere with the Pact mock.
+
+=== The contract
+Look in `target/pacts`.
+You should find a json contract.
+Inspect it to see how the contents of the `@Pact` method have been turned into json.
+
+What we want to be testing when we use Pact in our tests is the behaviour of our own code.
+Here, although we check the value of the hero fields, we're trying to make sure that the *fights service* is doing the right thing.
+Our expectation for the fights service is pretty low: it just has to pass through what it gets from the heroes service.
+
+[WARNING]
+This test is just on the edge of falling into a common Pact anti-pattern, testing the mock.
+We do *not* need to do detailed validation of what comes back from the pact server, because it's a mock that we define.
+
+A more interesting contract test would exercise some of the more complex fights functionality, while using the mock we've defined.
+Can you write one?
+(You might also need to define a villain pact.)
```

Comment:
... although we have that problem a little bit in the big one, too. Although less. 

I had an idea for that, by the way. Do you think there's a non-contrived transformation the Fights service could do? Like correcting the case of the fighter names, or ... um ... giving everyone WWF names by applying some rules to the name? 

A less silly transformation might be something like normalising the levels. I think we maybe had this conversation before and I had the exact same idea, but ... could the fight service normalise all the levels so that 10 was the max level for 90% of the fighters? I know some of the fighters have huge scores, so we'd want to divide by the 90th percentile, not the actual max. But it'd be a non-trivial data-based interaction. 

Or we could go back to the idea of having some selection criteria for the random fighters, so that the provider verification is testing more. 

In either case, I'll add the link. :)